tests/README.md

```diff
@@ -0,0 +1,120 @@
+# Honey-badger tests
+
+The `hbbft` crate comes with a toolkit for testing its various algorithms in simulated network environments.
+
+## Old vs new
+
+The old testing code can be found inside the `network` module and `.rs` files in the `tests` subdirectory that are not prefixed with `net_`. The newer networking code is contained inside the `net` module and the remaining `.rs` files.
+
+## VirtualNet
+
+Core of most tests is the `net::VirtualNet` struct, which simulates a network of nodes all running an instance of a distributed algorithm. Messages sent by these nodes are queued by the network. Every time the network is "cranked", a buffered message is delivered to its destination node and processed.
+
+Virtual networks can also host an adversary that can affect faulty nodes (which are tracked automatically) or reorder queued messages.
+
+To create a new network, the `NetBuilder` should be used:
+
+```rust
+// Create a network of 10 nodes, out of which 3 are faulty.
+let mut net = NetBuilder::new(0..10)
+    .num_faulty(3)
+    .using(move |node| { DynamicHoneyBadger::builder().build(node.netinfo) })
+    .build()
+    .expect("could not construct test network");
+```
+
+Algorithms that return a `Step` upon construction should use `using_step` instead.
+
+### Sending input
+
+`Input` can be sent to any node of the `VirtualNet` using the `send_input` method:
+
+```rust
+let input = ...;
+let _step = net.send_input(123, input).expect("algorithm failed");
+```
+
+While the resulting step is returned, it needn't be processed to keep the network going, as its messages are automatically added to the queue.
+
+Instead of targeting a node in particular, the same input can be sent to all nodes:
+
+```rust
+net.broadcast_input(input).expect("algorithm failed");
+```
+
+### Cranking the network
+
+The network advances through the `crank()` function, on every call
+
+1. the adversary is given a chance to re-order the message queue,
```

Comment:
Actually it seems that `Adversary::pre_crank` allows to mutate any part of the `VirtualNet`. So I think this sentence is not complete.

If you wanted to restrict the amount of state the adversary is allowed to mutate, I think it's better to do that by passing that exact amount of data to `pre_crank`. But having an all-powerful adversary has its positive sides as well. I'm fine with just updating this list item.

Reply:
I think that it is the responsibility of [the test implementer](https://github.com/poanetwork/hbbft/pull/216/files#diff-b5477a5a9bdae3f3f27bc239c3cab996R10) not to mutate the network in any way not allowed by our current model, maybe just note that "you may notice that it's possible for an adversary to mutate any part of `VirtualNet` (i.e. to change things beyond the scope of our adversary model), our current model only allows for an adversary to do: x, y, and z."

Reply:
> Actually it seems that Adversary::pre_crank allows to mutate any part of the VirtualNet. So I think this sentence is not complete.

Yes, that is the unfortunate reality. High on my list is adding a trait that restricts manipulation by the adversary to those operations they are actually allowed to perform, but I did not want to hold up review for this and added a note instead.

@DrPeterVanNostrand is right here. The `README.md` makes the mistake of assuming that one has already read all of the source code when reading the intro, which is silly. I will add more information to that specific section.